Backslash in paths generates C++ error

#### Summary:

I am using cmdStan in Julia in Windows 10. I apologize if this is a Julia related error. What happens is that when Stan generates the file for C++ it uses backslash in the file path, and this makes C++ throw an error.

#### Description:

I am not sure if this is an issue with cmdstan or stan.jl, so if I am opening the issue in the wrong place I apologize.

I am using Julia 1.0.0, cmdStan 2.20.0 and all packages on Julia are up to date. I am using Windows and did the whole work of using RTools to make cmdStan work.

I am using the Bernoulli example that is [in the documentation](https://stanjulia.github.io/Stan.jl/latest/WALKTHROUGH.html). It throws an error and the log says:

```
C:/Users/User/Documents/GitHub/Julia-Para-Economistas/tmp/bernoulli.hpp:21:37: error: incomplete universal character name \U
     reader.add_event(13, 11, "end", "C:\Users\User\Documents\GitHub\Julia-Para-Economistas\tmp\bernoulli.stan");
```

The issue is that for the paths, like `"C:\Users\User\Documents\GitHub\Julia-Para-Economistas\tmp\bernoulli.stan"`, stan is using backslash. Going in the bernoulli.hpp file and changing the backslash in the paths to slash solves the problem. 

So, the question is how do I make cmdStan make this change automatically? Is this an Julia issue or a cmdStan problem?


#### Reproducible Steps:

It is in Julia:

```
using CmdStan
using Mamba
using Cairo

set_cmdstan_home!("C:\\Users\\User\\Downloads\\cmdstan-2.20.0")

bernoullistanmodel = "
data { 
  int<lower=0> N; 
  int<lower=0,upper=1> y[N];
} 
parameters {
  real<lower=0,upper=1> theta;
} 
model {
  theta ~ beta(1,1);
    y ~ bernoulli(theta);
}"

stanmodel = Stanmodel(name="bernoulli", model=bernoullistanmodel)

bernoullidata = Dict("N" => 10, "y" => [0, 1, 0, 1, 0, 0, 0, 0, 0, 1])

rc, sim1 = stan(stanmodel, [bernoullidata], 
"C:\\Users\\User\\Documents\\GitHub\\Julia-Para-Economistas")
```

You can use Stan instead of CmdStan packages, both generate the same error

#### Current Output:

Julia throws _An error has occurred_. The log shows the following error:


```C:/Users/User/Documents/GitHub/Julia-Para-Economistas/tmp/bernoulli.hpp:21:37: error: incomplete universal character name \U
     reader.add_event(13, 11, "end", "C:\Users\User\Documents\GitHub\Julia-Para-Economistas\tmp\bernoulli.stan");
```

#### Expected Output:

Model should be estimated correctly, as it is when I change the paths by hand. 

#### Additional Information:

Maybe I did something wrong when I installed it? It took me a while to get cmdStan to find g++ in RTools so it could compile something. I am not sure if I did something wrong during installation and if it is necessary, I can describe every step I took during instalation. 

#### Current Version:
v2.20.0

Thanks in advance!

It looks like you're in roughly the right place - I think this is probably
a bug in the Stan compiler, though I'm surprised it's only surfacing now...
I'll look into it - there's a good chance it will be fixed when we switch
to stanc3.

On Tue, Aug 6, 2019 at 02:18 Daniel Coutinho <notifications@github.com>
wrote:

> Summary:
>
> I am using cmdStan in Julia in Windows 10. I apologize if this is a Julia
> related error. What happens is that when Stan generates the file for C++ it
> uses backslash in the file path, and this makes C++ throw an error.
> Description:
>
> I am not sure if this is an issue with cmdstan or stan.jl, so if I am
> opening the issue in the wrong place I apologize.
>
> I am using Julia 1.0.0, cmdStan 2.20.0 and all packages on Julia are up to
> date. I am using Windows and did the whole work of using RTools to make
> cmdStan work.
>
> I am using the Bernoulli example that is in the documentation. It throws
> an error and the log says:
>
> C:/Users/User/Documents/GitHub/Julia-Para-Economistas/tmp/bernoulli.hpp:21:37: error: incomplete universal character name \U
>      reader.add_event(13, 11, "end", "C:\Users\User\Documents\GitHub\Julia-Para-Economistas\tmp\bernoulli.stan");
>
> The issue is that for the paths, like
> "C:\Users\User\Documents\GitHub\Julia-Para-Economistas\tmp\bernoulli.stan",
> stan is using backslash. Going in the bernoulli.hpp file and changing the
> backslash in the paths to slash solves the problem.
>
> So, the question is how do I make cmdStan make this change automatically?
> Is this an Julia issue or a cmdStan problem?
> Reproducible Steps:
>
> It is in Julia:
>
> using CmdStan
> using Mamba
> using Cairo
>
> set_cmdstan_home!("C:\\Users\\User\\Downloads\\cmdstan-2.20.0")
>
> bernoullistanmodel = "
> data {
>   int<lower=0> N;
>   int<lower=0,upper=1> y[N];
> }
> parameters {
>   real<lower=0,upper=1> theta;
> }
> model {
>   theta ~ beta(1,1);
>     y ~ bernoulli(theta);
> }"
>
> stanmodel = Stanmodel(name="bernoulli", model=bernoullistanmodel)
>
> bernoullidata = Dict("N" => 10, "y" => [0, 1, 0, 1, 0, 0, 0, 0, 0, 1])
>
> rc, sim1 = stan(stanmodel, [bernoullidata],
> "C:\\Users\\User\\Documents\\GitHub\\Julia-Para-Economistas")
>
> You can use Stan instead of CmdStan, both generate the same error
> Current Output:
>
> Julia throws *An error has occurred*. The log shows the following error:
>
>      reader.add_event(13, 11, "end", "C:\Users\User\Documents\GitHub\Julia-Para-Economistas\tmp\bernoulli.stan");
>
> Expected Output:
>
> Model should be estimated correctly, as it is when I change the paths by
> hand.
> Additional Information:
>
> Maybe I did something wrong when I installed it? It took me a while to get
> cmdStan to find g++ in RTools so it could compile something. I am not sure
> if I did something wrong during installation and if it is necessary, I can
> describe every step I took during instalation.
> Current Version:
>
> v2.20.0
>
> Thanks in advance!
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/stan-dev/cmdstan/issues/720?email_source=notifications&email_token=AAGET3CXBJ6GLM5SELXTMHLQDDGORA5CNFSM4IJQBML2YY3PNVWWK3TUL52HS4DFUVEXG43VMWVGG33NNVSW45C7NFSM4HDQJPZQ>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAGET3AT3N6N7VJMMKPZGT3QDDGORANCNFSM4IJQBMLQ>
> .
>

--- a/make/program
+++ b/make/program
@@ -30,7 +30,7 @@ endif
 %.hpp : %.stan bin/stanc$(EXE)
 	@echo ''
 	@echo '--- Translating Stan model to C++ code ---'
-	$(WINE) bin/stanc$(EXE) $(STANCFLAGS) --o=$@ $<
+	$(WINE) bin/stanc$(EXE) $(STANCFLAGS) --o=$(subst  \,/,$@) $(subst  \,/,$<)
 
 
 .PRECIOUS: %.hpp

